### PR TITLE
Rewrite of DependentTestFinder

### PR DESCRIPTION
**Pull request by ReedOei**
Rewriting the Dependent Test Finder to use new methodology and dependency format.

### CODE REVIEW

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 1]
```diff
@@ -0,0 +1,600 @@
+/**
+ * Copyright 2014 University of Washington. All Rights Reserved.
+ * @author Wing Lam
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: You can use your name here, Reed :)

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+        public List<String> getTestsBefore(final String testName) {
+            return testOrder.subList(0, testOrder.indexOf(testName));
+        }
+    }
+
+    private static TestExecResult runTestOrder(final List<String> order) {
+        return new FixedOrderRunner(order).run().getExecutionRecords().get(0);
+    }
+
+    public static void main(String[] args) {
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: You can remove this main method. We have no intention of letting the class be used this way at least for now.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+    }
+
+    private List<String> generatePrimeOrder(final List<String> originalOrder,
+                                            final List<String> newOrder) {
+        final List<String> primeOrder = new ArrayList<>(newOrder);
+
+        final int origIndex = originalOrder.indexOf(dependentTestName);
+        final int primeIndex = primeOrder.indexOf(dependentTestName);
+
+        // TODO: Make sure no tests are in order twice.
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: Please finish this TODO.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+
+    public static void writeToFile(final List<ParallelDependentTestFinder> dtFinders) {
+        writeToFile(dtFinders, DT_FILE);
+    }
+
+    /**
+     * Writes all the dependency information to a file.
+     * @param dtFinders The dtFinders that contain the dependency information we need.
+     */
+    public static void writeToFile(final List<ParallelDependentTestFinder> dtFinders, final File outputFile) {
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: Instead of writing your own method to convert the precomputed dependencies to a string, please use the test data class here instead. https://github.com/winglam/dependent-tests-impact/pull/22/files/ba79a576feb562d4c089c3042b0e2f927a1f626e#diff-303b5f1678f338b5174bbfbdbfeef0b7R11

I will add a comment to Henry to delete his class after you have merged your PR.

You should also replace all uses of Strings to represent the precomputed dependencies with this test data class.

Comment (ReedOei) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: I'm not sure how to incorporate this into my current program. Right now, I use a map to keep track of dependencies (each test maps to the tests that must be run before it). If I use this data class, both the test and it's dependencies will be stored in the class, so I won't have a separate key/value for the map. It isn't strictly necessary to use the map, but it seems like the most efficient option.

Additionally, that class doesn't seem to contain the correct information--are you suggesting I write my own version that has what I need (hence the "comment to Henry to delete his class...")?

Comment (winglam) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: You can use the data class and a map by having the name of the dependent test be mapped to a set of the data class.

The class indeed does not contain the correct information. The only information that I think you would need is just what I added as a comment to Henry's PR.

BTW, you mention that you are only keeping track of which tests need to run before a dependent test. If that is true, double check your work to ensure that if a test suite containing T1 and T2 and the only dependency is that T2 must always come AFTER T1, then your work here will still terminate and output the correct result (one dependence saying T1 needs to come before T2). You may also want to imagine a slightly more difficult case and see whether your work will be able to handle such a case.

Comment (ReedOei) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: I had been keeping track of which tests need to be run after by adding another entry to the map (saying that T1 has to come before T2, rather than T2 has to come after T1). I've just realized that this approach is probably not the best behavior, as it adds dependencies that aren't there.

In addition to the change you suggested on Henry's pull request, I think that I will have to add another field, either to specify whether the independent test needs to come before or after the dependent test, or include both a before test and after test in the data class. I think the best way to do this might be to make the data class hold a set of the before tests and a set of the after tests, then I could simply map each dependent test to a single instance of the data class.

Additionally, I think that we can omit the "when executed after: []" line that comes after the "revealed different behavior..." line. This is because the dependent test should be revealed any time one of the dependencies (either a test coming before or after) is not satisfied, right?

Comment (winglam) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: "(saying that T1 has to come before T2, rather than T2 has to come after T1). I've just realized that this approach is probably not the best behavior, as it adds dependencies that aren't there."
I agree. I think the test data class should be improved to keep track of whether the dependence is a before or after dependence. 

"I think the best way to do this might be to make the data class hold a set of the before tests and a set of the after tests, then I could simply map each dependent test to a single instance of the data class."
Including a field for both and making them be sets sounds good to me.

"Additionally, I think that we can omit the "when executed after: []" line that comes after the "revealed different behavior..." line."
My suggestion would be that the test data class contains a set of before and a set of after tests but your map of precomputed dependences still maps from a dependent test to a set of test data classes. The reason for this implementation is that we can actually generate a "debug" log from this implementation. 

You are correct that the final product of the precomputed dependencies list would only need for each dependent test, which tests need to come before and after it. However, with just that information we lose the reason as to how we got those before and after tests and more importantly, what could have happened without them.

For example, the "debug" log might tell us something like T5 needs to come after T4 otherwise T4 will pass instead of fail compared to the original order. Another dependence of T4 might say T3 needs to come before T4 otherwise T4 will error instead of fail compared to the original order. These are two separate dependencies that are both necessary for T4 to have the same behavior as the original order but they affect T4 differently and with the debug log we can identify these cases. The debug log will also clearly explain to us why T4 needs T3 to come before it and T5 to come after as opposed to just knowing that T4 won't have the same behavior if those two tests are not in the right order.

To sum up, I would expect the internal representation to support the idea of the debug log and the tool to output the debug log for those that are curious about how the precomputed dependencies are generated. However, our tool should also generate a concise precomputed dependencies list where each dependent test only shows up once and the only fields needed are before and after tests. For the debug log, each dependent test may appear more than once and you would need before tests, after tests, the dependent test's behavior without those before and after tests, the test's behavior in the original order, and the test list that revealed that dependent tests behavior.

If you want more clarification, feel free to shoot me a text and come by the office.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/TestData.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        this.beforeTests = beforeTests;
+        this.afterTests = afterTests;
+        this.revealed = revealed;
+        this.revealingOrder = revealingOrder;
+    }
+
+    @Override
+    public String toString() {
+        // If the intended behavior is the revealed behavior, then the before/after tests are the
+        // tests needs before/after to make it have the intended behavior.
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/TestData.java]: Can you elaborate under what scenario would the intended behavior be the same as the revealed behavior?

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+
+        dependentTestResult = this.originalOrder.getResult(dependentTestName);
+        debugLog = new ArrayList<>();
+    }
+
+    public ParallelDependentTestFinder(final String dependentTestName,
+                                       final List<String> originalOrder,
+                                       final List<String> newOrder,
+                                       final List<String> filesToDelete,
+                                       final Map<String, TestData> knownDependencies) {
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: Perhaps you are still in the middle of changing this implementation but if not, I think the knownDependencies here should be a Map<String, Set<TestData>>. See below and my comment on your writeToFile method earlier today for my explanation why.

"My suggestion would be that the test data class contains a set of before and a set of after tests but your map of precomputed dependences still maps from a dependent test to a set of test data classes. The reason for this implementation is that we can actually generate a "debug" log from this implementation."

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+                .map(ParallelDependentTestFinder::toString)
+                .reduce("", String::concat);
+    }
+
+    @Override
+    public String toString() {
+        final TestData testData = knownDependencies.get(dependentTestName);
+
+        if (testData != null) {
+            return testData.toString();
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: Instead of returning the testData as a string, Henry's implementation should have a way to get the testData object. That way Henry can then update the precomputed dependencies' Map<String, <Set<TestData>> with the new TestData object or you could do add in this class and just return to Henry the knownDependencies (but I don't think you are currently doing that either, feel free to ignore this comment if you are).

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+            FileWriter output = null;
+            BufferedWriter writer = null;
+            try {
+                output = new FileWriter(outputFile);
+                writer = new BufferedWriter(output);
+
+                for (final ParallelDependentTestFinder dtFinder : dtFinders) {
+                    if (includeDebugLog) {
+                        writer.write(dtFinder.debugLogToString() + "\n");
+                    }
```
Comment (winglam) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java]: The debug log and the actual printing of the precomputed dependencies should be separate. The precomputed dependencies only cares about what the dependent test is and what tests it depends on to come before and after it. The debug log should have all fields of the TestData objects be printed. 

Its fine to have a method that does both but the debug log does not need the info of the precomputed dependencies (it should be a else here) and this method should reside in the Runner class and take a set of TestData objects.

**Comment by ReedOei** [comment]
I have made the changes to the ParallelDependentTestFinder that make it return the map containing the set of TestData objects. I think it is ready to be merged now so Henry can integrate it into his tool.

**Review by winglam — APPROVED** [review]
LGTM. For the code you had before where you wrote the Map of precomputed dependencies to a file, you should send that code to Henry so that he can reuse it if it makes sense for him to do so.
